src/rpp/rpp/subjects/replay_subject.hpp

```diff
@@ -0,0 +1,158 @@
+//                   ReactivePlusPlus library
+//
+//           Copyright Aleksey Loginov 2022 - present.
+//  Distributed under the Boost Software License, Version 1.0.
+//     (See accompanying file LICENSE_1_0.txt or copy at
+//           https://www.boost.org/LICENSE_1_0.txt)
+//
+//  Project home: https://github.com/victimsnino/ReactivePlusPlus
+
+#pragma once
+
+#include <rpp/schedulers/fwd.hpp>
+#include <rpp/subjects/fwd.hpp>
+
+#include <rpp/observers/observer.hpp>
+#include <rpp/subjects/details/base_subject.hpp>
+#include <rpp/subjects/details/subject_state.hpp>
+
+#include <list>
+#include <optional>
+
+namespace rpp::subjects::details
+{
+template<rpp::constraint::decayed_type Type>
+class replay_strategy
+{
+    struct replay_state
+    {
+        std::optional<size_t>                    count;
+        std::optional<rpp::schedulers::duration> duration;
+
+        std::list<Type>                                    values{};
+        std::list<rpp::schedulers::clock_type::time_point> time_points{};
+
+        std::mutex          mutex{};
+        subject_state<Type> state{};
+    };
+
+    struct observer_strategy
+    {
+        std::shared_ptr<replay_state> state;
+
+        void set_upstream(const disposable_wrapper& d) const noexcept { state->state.add(d); }
+
+        bool is_disposed() const noexcept
+        {
+            return state->state.is_disposed();
+        }
+
+        void on_next(const Type& v) const
+        {
+            std::unique_lock lock{state->mutex};
+            if (state->count.has_value())
+            {
+                if (state->values.size() == state->count.value())
+                {
+                    state->values.pop_front();
+                    if (state->duration.has_value())
+                    {
+                        state->time_points.pop_front();
+                    }
+                }
+            }
+            if (state->duration.has_value())
+            {
+                auto now = rpp::schedulers::clock_type::now();
+                while (!state->time_points.empty() && (now - state->time_points.front() > state->duration.value()))
+                {
+                    state->values.pop_front();
+                    state->time_points.pop_front();
+                }
+                state->time_points.push_back(now);
+            }
+
+            state->values.push_back(v);
+            state->state.on_next(v);
+        }
+
+        void on_error(const std::exception_ptr& err) const
+        {
+            std::unique_lock lock{state->mutex};
```

Comment:
Why we should lock mutex there? mutex is guarding only lists, isn't it?

Reply:
Depends do we want to make our replay_subject serialized too while we are at it?

Reply:
To be honest, i don't know. I think, we should have replay + serialized_replay subjects in this case?

Reply:
done, added both